### DISCUSSION TITLE
Nested provider and atom value resolution: allow derive store?

### DISCUSSION THREAD

**Discussion by yf-yang** [pmndrs/jotai#1894] [Ideas] [answered]
I'll start with an example:
``` typescript
export const countAtom = atom(0);

function Counter() {
  const [count, setCount] = useAtom(countAtom);
  return (
    <>
      <h1>{count} </h1>
      <button onClick={() => setCount((value) => value + 1)}>one up</button>
    </>
  );
}

const store = createStore();
store.set(countAtom, 100);
const App = () => (
  <>
    <Counter />
    <Provider store={store}>
      <Counter />
      <Provider>
        <Counter />
      </Provider>
    </Provider>
  </>
);
```
Here I'll call those three counters from up to bottom 1/2/3. Counter 2 is initialized with 100, while counter 3 isinitialized with 0.
I checked the source code, so I can understand the underlying reason. I am not sure if similar discussions exist, if so, please redirect.

# The Problem
1. Counter3 cannot get the value of Counter2.
2. When Counter3 wants to update `countAtom` of Counter2, it could never do so.

# Why it is a problem
I'll illustate with a business scenario. Suppose there is an organization where people are organized as groups, then groups form a department, then department form a Business Unit, so for a app, it could have four components: BU、department、group and person. Each component has some metadata that is shared across all of its child components, and each component is deeply nested, so I would not choose useState and pass it through the tree, instead I will create a Provider for BU、department and group, then access BU metadata/department metadata/group metadata from the person component.

Since each BU could have multiple department components (so do departments and groups), those providers are necessary. It should be considered as something whose reusability is between pure global and pure local. For multiple person component, the department metadata is just like global one, because the tree is really deep. However, for BU, it is undoubtly local. I cannot put department metadata in BU component, because one BU has multiple departments.

I am not sure if this is the best practice, but multiple layer of Provider is a solution. If so, then those problems would become painful.

# Idea
I can pass `store` to the inner provider, then the inner provider actually makes no use, I can simply remove it. That is not what I expect. Actually, sometimes I would appreciate the benefits of isolating stores between nested providers, but sometimes I also want to pass some atoms to the inner provider.
Therefore, one possible solution is adding an API so that the developer can explicitly ask a new store to inherit some of those atoms from an existing store when the new store is initialized.
``` typescript
const storeOuter = createStore();
const storeInner = createStore(storeOuter, [countAtom]);
store.set(countAtom, 100);
const App = () => (
  <>
    <Counter />
    <Provider store={storeOuter}>
      <Counter />
      <Provider store={storeInner}>
        <Counter />
      </Provider>
    </Provider>
  </>
);
```
For `countAtom`, the inner provider is transparent, but for other atoms I use inside the inner provider, they are isolated from their values in the outer provider and the default store.
Signature:
``` typescript
// from === undefined => create a brand new one
// atoms === undefiend => inherit all existing atoms
function createStore(from?: Store, atoms?: Atom[]);
```

# Some implementation considerations
- A store should store a reference of its parent stores. In `readAtomState`, if `getAtomState` fails to get one, search its parent stores.
- Maybe multi-inheritance is needed, maybe it is not. I am not sure if all the multi-inheritance scenarios could be refactored.
- In this way, store inheritance order may not follow provider nesting order. If one do so, it could be confusing, guidance or runtime check is needed.

**Comment by yf-yang** [comment]
May be duplicate of https://github.com/pmndrs/jotai/discussions/1601, let me take a look

**Comment by yf-yang** [comment]
I am confused, it seems jotai-molecules cannot resolve this issue. Let me clarify the problem:
When using React context, I can declare multiple contexts, so each one responsible for its own value (BU context, department context, group department). However, in jotai, there is exactly one anonymous context (wrapped inside Provider), so when I create a nested Provider, it shadows the outer Provider (and the outer context). Even I know that department Provider is only responsible for department metadata atom, I still can't get the outer BU metadata atom. Is this problem stated somewhere?

**Reply by yf-yang** [reply]
In this scerario, the department provider's context should come from two sources: one is the newly created context (for department metadata), one is from the BU context. An API to combine them is needed.

**Comment by yf-yang** [comment, marked as answer]
jotai-molecules does solve the problem.

For whom is concerned, check its `useScope` implmentation, it handles inheritance

``` typescript
const scope1 = createScope(0);
const scope2 = createScope(10);

const molecule1 = molecule((_getMol, getScope) => {
  const initialCont = getScope(scope1);
  return atom(initialCont);
});
const molecule2 = molecule((_getMol, getScope) => {
  const initialCont = getScope(scope2);
  return atom(initialCont);
});

function Counter() {
  const [count1, setCount1] = useAtom(useMolecule(molecule1));
  const [count2, setCount2] = useAtom(useMolecule(molecule2));
  return (
    <>
      <h1>{count1} </h1>
      <h1>{count2} </h1>
      <button onClick={() => setCount1((value) => value + 1)}>
        one up upper
      </button>
      <button onClick={() => setCount2((value) => value + 1)}>
        one up lower
      </button>
    </>
  );
}

const App = () => (
  <>
    <Counter />
    <ScopeProvider scope={scope1} value={100}>
      <Counter />
      <ScopeProvider scope={scope2} value={110}>
        <Counter />
      </ScopeProvider>
    </ScopeProvider>
  </>
);
```

**Comment by dai-shi** [comment]
I just developed another util for this kind of use cases. Hope someone can try.
https://github.com/jotaijs/jotai-scope/pull/2
